<?xml version="1.0"?>
<robot name="sensor_gps_with_dropout" xmlns:xacro="http://ros.org/wiki/xacro">
     <xacro:include filename="$(find gazebo_sensor_collection)/urdf/includes/materials.urdf.xacro"/>

     <!-- ====================================================
     Helpers
     ========================================================= -->
     <xacro:macro name="plot" params="plot_id plot_range_index plot_crossrange_index x y z rotation row_count row_width row_length average_height average_density">
          <plot_id>${plot_id}</plot_id>
          <plot_range_index>${plot_range_index}</plot_range_index>
          <plot_crossrange_index>${plot_crossrange_index}</plot_crossrange_index>
          <x>${x}</x>
          <y>${y}</y>
          <z>${z}</z>
          <rotation>${rotation}</rotation>
          <row_count>${row_count}</row_count>
          <row_width>${row_width}</row_width>
          <row_length>${row_length}</row_length>
          <average_height>${average_height}</average_height>
          <average_density>${plot_id}</average_density>
     </xacro:macro>
     <xacro:macro name="periodic_noise_box" params="cx cy cz w l h rot nps nmpm nmpv nvpm nvpv">
          <cx>${cx}</cx>
          <cy>${cy}</cy>
          <cz>${cz}</cz>
          <w>${w}</w>
          <l>${l}</l>
          <h>${h}</h>
          <rot>${rot}</rot>
          <nps>${nps}</nps>
          <nmpm>${nmpm}</nmpm>
          <nmpv>${nmpv}</nmpv>
          <nvpm>${nvpm}</nvpm>
          <nvpv>${nvpv}</nvpv>
     </xacro:macro>

     <!-- ====================================================
     GPS
     ========================================================= -->
     <xacro:macro name="sensor_gps_with_dropout" params="name parent *origin color ref_lat ref_long ref_head drift vel_drift noise vel_noise update_rate">

          <joint name="${name}_joint" type="fixed">
               <axis xyz="0 1 0"/>
               <xacro:insert_block name="origin"/>
               <parent link="${parent}"/>
               <child link="${name}_link"/>
          </joint>

          <link name="${name}_link">
               <inertial>
                    <mass value="0.001"/>
                    <origin rpy="0 0 0" xyz="0 0 0"/>
                    <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.000001" iyz="0" izz="0.0001"/>
               </inertial>
               <visual>
                    <origin rpy="0 0 0" xyz="0 0 0"/>
                    <geometry>
                         <mesh filename="file://$(find gazebo_sensor_collection)/meshes/gps/antenna_3GO16.stl" scale="1.0 1.0 1.0"/>
                    </geometry>
                    <xacro:color_material color="${color}"/>
               </visual>
               <collision>
                    <origin rpy="0 0 0" xyz="0 0 0"/>
                    <geometry>
                         <mesh filename="file://$(find gazebo_sensor_collection)/meshes/gps/antenna_3GO16.stl" scale="1.0 1.0 1.0"/>
                    </geometry>
               </collision>
          </link>

          <!-- ====================================================
          Gazebo
          ========================================================= -->

          <xacro:color_gazebo parent="${name}_link" color="${color}"/>

          <gazebo>
               <plugin name="${name}_controller" filename="libgazebo_ros_gps_with_dropout.so">
                    <xacro:include filename="$(find gazebo_sensor_collection)/config/gps/custom_gps_behavior.geofence"/>
                    <xacro:include filename="$(find gazebo_sensor_collection)/config/gps/plots.xacro"/>
                    <alwaysOn>1</alwaysOn>
                    <updateRate>${update_rate}</updateRate>
                    <bodyName>${name}_link</bodyName>
                    <frameId>${name}_link</frameId>
                    <topicName>/${name}/fix</topicName>
                    <velocityTopicName>/${name}/fix_velocity</velocityTopicName>
                    <dataTopicName>/${name}</dataTopicName>
                    <referenceLatitude>${ref_lat}</referenceLatitude>
                    <referenceLongitude>${ref_long}</referenceLongitude>
                    <referenceHeading>${ref_head}</referenceHeading>
                    <drift>${drift} ${drift} ${drift}</drift>
                    <gaussianNoise>${noise} ${noise} ${noise}</gaussianNoise>
                    <velocityDrift>${vel_drift} ${vel_drift} ${vel_drift}</velocityDrift>
                    <velocityGaussianNoise>${vel_noise} ${vel_noise} ${vel_noise}</velocityGaussianNoise>
               </plugin>
          </gazebo>
     </xacro:macro>

     <!--<xacro:include filename="$(find gazebo_sensor_collection)/config/gps/dropouts.xml"/>-->

</robot>
